### ISSUE TITLE
support kwargs for params without having to create a Dict

### ISSUE THREAD

**Issue by oxinabox** [JuliaCloud/AWS.jl#341]
Every function in the API takes a `param::Dict{Any,String}`

It would be useful if we could also accept normal julia keyword arguments
and have that get wrapped up into a `param` Dict automatically.
via `params = Dict{String, Any}(string(k)=>v for (k,v) in pairs(kwargs))`

**Comment by mattBrzezinski** [comment]
Yes, this would be nice to do! One annoying thing is the usage of hyphens in optional parameters for instance `S3.list_buckets()` has a `max-keys` optional parameter (among others).

We would need to introduce some check to convert `_` to `-`, but otherwise this is very doable.

**Comment by mattBrzezinski** [comment]
I'm looking into this now, and this won't be easily possible to do. The issue stems from certain services having argument names with hyphens, or underscores, or both.

For example, `SSO` has an argument called `x-amz-sso_bearer_token`. If we want this to be a kwarg we could call it `x_amz_sso_bearer_token`, but we then need  to convert this back to `x-amz-sso_bearer_token`, but its' not as simple as just converting underscores to hyphens as we only want to do this conversion for some.

We would essentially need to create some master list of `expected_name => valid_julia_symbol`, which seems like a lot of work.

AWS' inconsistencies strike again!

**Comment by oxinabox** [comment]
What if we translated `-` into `_`, and `_` into `__` (double underscore)?

**Comment by mattBrzezinski** [comment]
> What if we translated `-` into `_`, and `_` into `__` (double underscore)?

That works! I don’t really like going down this gross path but I suppose it’s the only real option.

**Comment by oxinabox** [comment]
I think `var"foo-bar_qux"` might work for kwarg names but I haven't checked

**Comment by omus** [comment]
How about this as an alternative? Use varargs to make an interface that is keyword-like but avoids having to mess with translation:

```julia
julia> f(aws_kwargs::Pair{String,<:Any}...) = aws_kwargs
f (generic function with 1 method)

julia> f("x-amz-sso_bearer_token" => "1234")
("x-amz-sso_bearer_token" => "1234",)
```

I think this is pretty reasonable from a user interface perspective and avoids having to learn the special translation logic when reading AWS documentation. The only real downside here is a little bit of processing required to be done.

**Comment by mattBrzezinski** [comment]
> How about this as an alternative? Use varargs to make an interface that is keyword-like but avoids having to mess with translation:
> 
> ```julia
> julia> f(aws_kwargs::Pair{String,<:Any}...) = Dict(aws_kwargs)
> f (generic function with 1 method)
> 
> julia> f("x-amz-sso_bearer_token" => "1234")
> Dict{String, String} with 1 entry:
>   "x-amz-sso_bearer_token" => "1234"
> ```
> 
> I think this is pretty reasonable from a user interface perspective and avoids having to learn the special translation logic when reading AWS documentation. The only real downside here is a little bit of processing required to be done.

It's late, I'm tired, so I might be missing something here. In the example above we would want `aws_kwargs` to actually be a `kwargs...`, since all these values should be optional arguments to the request we make. An example,

```julia
combine(; kwargs...) = Dict{String, Any}(string(k)=>v for (k,v) in pairs(kwargs))

function S3.list_objects(bucket; aws_config=global_aws_config(), kwargs...)
    return s3("GET", "$(bucket)", combine(kwargs...); aws_config=aws_config)
end
```

This way the user can just do `? S3.list_objects`, see whatever optional arguments there are and pass them along as,
```julia
S3.list_objects("my-bucket"; max-keys=1)
```

But obviously that won't work :/ since there's a hyphen in `max-keys`. We could do something with the `InputShape` for each endpoint, but briefly thinking about it, we probably wouldn't get much closer to having something nice.

**Comment by omus** [comment]
> In the example above we would want `aws_kwargs` to actually be a `kwargs...`, since all these values should be optional arguments to the request we make.

My alternative is suggesting forgoing using Julia keyword arguments for something keyword-like in order to avoid having to conform to Julia's requirements for variable names. For your posted example:

```julia
function S3.list_objects(bucket, aws_kwargs::Pair{String,<:Any}...; aws_config=global_aws_config())
    return s3("GET", "$(bucket)", Dict(aws_kwargs...); aws_config=aws_config)
end
```
And then call it with:
```julia
S3.list_objects("my-bucket", "max-keys" => 1)
```
The advantage I for this approach is no translation is required to go from Julia keyword to AWS optional argument.

> This way the user can just do `? S3.list_objects`, see whatever optional arguments there are and pass them along

You should also be able to document optional arguments as part of the docstring using this proposal.

**Comment by oxinabox** [comment]
> I think `var"foo-bar_qux"` might work for kwarg names but I haven't checked.


Confirmed this does work
```julia
julia> bar(;kwargs...) = kwargs
bar (generic function with 1 method)

julia> bar(var"ab-cd_ef"=2)
pairs(::NamedTuple) with 1 entry:
  Symbol("ab-cd_ef") => 2
  
 julia> qux(; var"ab-cd_ef"=nothing) = var"ab-cd_ef"
qux (generic function with 1 method)

julia> qux(; var"ab-cd_ef"=42)
42
```

**Comment by mattBrzezinski** [comment]
Ah I see thanks for clarifying! I'm a bit hesitant to put in the work to change from a `Dict` to a `Pair` as they're both pretty similar? Idk if you have thoughts on this @oxinabox?

The `var"ab-cd_ef"` solution could work, and I feel like this is in a bit of a better direction? But I feel like it would still be a bit awkward for the end user when they're doing something like:

```julia
using AWS
@service S3

resp = S3.list_bucket("my-bucket"; var"max-keys"=1, delimiter="/", prefix="pref")
```

Although, maybe this is a better step forward? At least if you know an optional argument contains a hyphen you need to preface it with `var"optional-arg"`?

**Comment by omus** [comment]
If you really, *really* want to use keywords then you could make use of a macro to allow the use of `-` as part of a keyword. Here is some quick code to show that the macro can capture the hyphenated keyword as a single expression:
```julia
macro foo(exprs...)
           @show exprs
           :(nothing)
       end
@foo (macro with 1 method)

julia> @foo(; max-keys=1, ab-cd_ef-gh=2)
exprs = (:($(Expr(:parameters, :($(Expr(:kw, :(max - keys), quote
    #= REPL[4]:1 =#
    1
end))), :($(Expr(:kw, :((ab - cd_ef) - gh), quote
    #= REPL[4]:1 =#
    2
end)))))),)
```
There needs to be some additional design here like deciding if you have `S3.@list_bucket` or have a generic macro that can be used with any AWS call `@aws_kwarg S3.list_bucket(...; max-keys=1)`

**Comment by mattBrzezinski** [comment]
Personally I think for an endpoint required arguments should be positional, and any optional ones should be keywords. Not sure what others think of this, or what their opinions are. I might not be the best use case for determining this, and maybe other users of the package should provide input.

Seems like there are a few options here,

```julia
resp = S3.list_objects("my-bucket", Dict("max-keys"=>1))
```

```julia
resp = S3.list_objects("my-bucket", ("max-keys"=>1))
```

```julia
resp = S3.list_objects("my-bucket"; var"max-keys"=1)
```

```julia
resp = @kw S3.list_objects("my-bucket"; max-keys=1)
```

```julia
resp = S3.@list_objects("my-bucket"; max-keys=1)
```

**Comment by oxinabox** [comment]
Can we pull stats on how common `_` and `-` are in arguments? Ideally broken down by service (since tbh i don't care about what is common in GroundControl, but i definately do for CloudWatch)

 - If `_` is rare and `-` common, I lean towards the translate `-` to `_` and `__` to `-`. (since you would rarely need to remember the rule beyond `-` <=>`_`)
 - if `_` is common and `-` is rare, I lean towards the `var"ab_cd-ef"` solution. (Since you would rarely need to use `var` since most things wouldn't have `-`)
 - if both rare then I lean still lean towards  `var"ab_cd-ef"`  solution
 - if both common then I lean towards the `Vararg{Pair{String}}` solution. (it's a bit less nice than being able to use kwargs, but its pretty good)

**Comment by mattBrzezinski** [comment]
Yep I can pull numbers down, just off-hand looking at it, seems like `-` is rare, only used in `S3` and for various header specific parameters. `_` is used throughout various services quite often.

**Comment by mattBrzezinski** [comment]
All services containing parameters with either a hyphen, underscore, or both in them. Note, SSO is the only service which has both.

```json
[
  {
    "clouddirectory-2017-01-11.normal.json": {
      "-": 64,
      "both": 0,
      "_": 0
    }
  },
  {
    "mobile-2017-07-01.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "chime-2018-05-01.normal.json": {
      "-": 100,
      "both": 0,
      "_": 0
    }
  },
  {
    "runtime.lex-2016-11-28.normal.json": {
      "-": 32,
      "both": 0,
      "_": 0
    }
  },
  {
    "cloudsearchdomain-2013-01-01.normal.json": {
      "-": 1,
      "both": 0,
      "_": 0
    }
  },
  {
    "mq-2017-11-27.normal.json": {
      "-": 17,
      "both": 0,
      "_": 0
    }
  },
  {
    "mobileanalytics-2014-06-05.normal.json": {
      "-": 2,
      "both": 0,
      "_": 0
    }
  },
  {
    "qldb-2019-01-02.normal.json": {
      "-": 0,
      "both": 0,
      "_": 8
    }
  },
  {
    "greengrass-2017-06-07.normal.json": {
      "-": 24,
      "both": 0,
      "_": 0
    }
  },
  {
    "location-2020-11-19.normal.json": {
      "-": 4,
      "both": 0,
      "_": 0
    }
  },
  {
    "quicksight-2018-04-01.normal.json": {
      "-": 59,
      "both": 0,
      "_": 0
    }
  },
  {
    "glacier-2012-06-01.normal.json": {
      "-": 21,
      "both": 0,
      "_": 0
    }
  },
  {
    "lookoutvision-2020-11-20.normal.json": {
      "-": 10,
      "both": 0,
      "_": 0
    }
  },
  {
    "polly-2016-06-10.normal.json": {
      "-": 2,
      "both": 0,
      "_": 0
    }
  },
  {
    "runtime.sagemaker-2017-05-13.normal.json": {
      "-": 9,
      "both": 0,
      "_": 0
    }
  },
  {
    "appconfig-2019-10-09.normal.json": {
      "-": 8,
      "both": 0,
      "_": 15
    }
  },
  {
    "mediapackage-vod-2018-11-07.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "customer-profiles-2020-08-15.normal.json": {
      "-": 16,
      "both": 0,
      "_": 0
    }
  },
  {
    "codeguruprofiler-2019-07-18.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "nimble-2020-08-01.normal.json": {
      "-": 22,
      "both": 0,
      "_": 0
    }
  },
  {
    "lambda-2015-03-31.normal.json": {
      "-": 6,
      "both": 0,
      "_": 0
    }
  },
  {
    "apigateway-2015-07-09.normal.json": {
      "-": 4,
      "both": 0,
      "_": 176
    }
  },
  {
    "lex-models-2017-04-19.normal.json": {
      "-": 0,
      "both": 0,
      "_": 2
    }
  },
  {
    "dataexchange-2017-07-25.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "amplifybackend-2020-08-11.normal.json": {
      "-": 0,
      "both": 0,
      "_": 6
    }
  },
  {
    "connectparticipant-2018-09-07.normal.json": {
      "-": 8,
      "both": 0,
      "_": 0
    }
  },
  {
    "kinesis-video-archived-media-2017-09-30.normal.json": {
      "-": 2,
      "both": 0,
      "_": 0
    }
  },
  {
    "medialive-2017-10-14.normal.json": {
      "-": 6,
      "both": 0,
      "_": 0
    }
  },
  {
    "apigatewayv2-2018-11-29.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "cloudfront-2020-05-31.normal.json": {
      "-": 23,
      "both": 0,
      "_": 0
    }
  },
  {
    "s3control-2018-08-20.normal.json": {
      "-": 60,
      "both": 0,
      "_": 0
    }
  },
  {
    "iot-2015-05-28.normal.json": {
      "-": 8,
      "both": 0,
      "_": 0
    }
  },
  {
    "cognito-sync-2014-06-30.normal.json": {
      "-": 1,
      "both": 0,
      "_": 0
    }
  },
  {
    "iot1click-devices-2018-05-14.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "pinpoint-2016-12-01.normal.json": {
      "-": 178,
      "both": 0,
      "_": 0
    }
  },
  {
    "mediastore-data-2017-09-01.normal.json": {
      "-": 13,
      "both": 0,
      "_": 0
    }
  },
  {
    "kinesis-video-media-2017-09-30.normal.json": {
      "-": 1,
      "both": 0,
      "_": 0
    }
  },
  {
    "runtime.lex.v2-2020-08-07.normal.json": {
      "-": 17,
      "both": 0,
      "_": 0
    }
  },
  {
    "schemas-2019-12-02.normal.json": {
      "-": 3,
      "both": 0,
      "_": 0
    }
  },
  {
    "s3-2006-03-01.normal.json": {
      "-": 467,
      "both": 0,
      "_": 0
    }
  },
  {
    "mediapackage-2017-10-12.normal.json": {
      "-": 3,
      "both": 0,
      "_": 1
    }
  },
  {
    "ebs-2019-11-02.normal.json": {
      "-": 13,
      "both": 0,
      "_": 0
    }
  },
  {
    "sso-2019-06-10.normal.json": {
      "-": 0,
      "both": 4,
      "_": 7
    }
  },
  {
    "codeartifact-2018-09-22.normal.json": {
      "-": 51,
      "both": 0,
      "_": 0
    }
  }
]

```

**Comment by oxinabox** [comment]
I think we are in this case:

> If _ is rare and - common, I lean towards the translate - to _ and __ to -. (since you would rarely need to remember the rule beyond - <=>_)

It seems `_` is rare at all, and never used in a device I care about.
Unless API Gateway is a super important service, and needs you to constantly configure things with those optional arguments.

**Comment by omus** [comment]
Although unlikely do check for any params named `aws-config` that could conflict with the AWS.jl `aws_config`. 

Aside: I'll note I kind of always felt that the `aws_config` should be the optional first argument for the functions (e.g. define `f(::AWSConfig, args...)`  and `f(args...)`)

**Comment by ericphanson** [comment]
just to say I like the optional pairs as trailing position arguments (https://github.com/JuliaCloud/AWS.jl/issues/341#issuecomment-832704581). JuMP uses this pattern too for passing options to solvers: https://jump.dev/JuMP.jl/v0.21.10/reference/models/#JuMP.set_optimizer_attributes (which is very similar in that the names are defined by external binaries most the time).

**Comment by omus** [comment]
I'll mention that boto3 uses CamelCase for its keyword arguments. For [example `list_objects_v2` takes `MaxKeys`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_objects_v2) as an argument. We may also want to adopt this format as the response to `list_object_v2` calls includes "MaxKeys". It would be nice for these to be consistent.

**Comment by iamed2** [comment]
Boto3 gets its argument names from the API files: https://github.com/aws/aws-sdk-js/blob/a42ad578686801607b88358b7a85da48a8952bdc/apis/s3-2006-03-01.normal.json#L6516

I think we probably should too.

**Comment by iamed2** [comment]
> I believe that our low-level interface should not perform any parameter translation

What do you mean by this? What do you consider to be the canonical parameter name?

As discussed above, AWS has the name for a parameter (e.g., `MaxKeys`, `RequestPayer`), and then the location of the parameter (e.g., `max-keys` in the `querystring`, `x-amz-request-payer` in the `header`). Could you provide an example like the one you did for `CloudFormation` (currently this is typed `cloudformation`, probably best to use that to avoid confusion with the higher-level interface) but for `s3`?

**Comment by omus** [comment]
I'm looking into this again and it appears that the switch to using the Smithy based service definitions files on the v2 branch have standardized parameters names. We can use theses directly safely as keywords or alternatively convert them into snake_case for Julia consistency. As part of this it makes sense to make `aws_config` an optional positional parameter.
